osu.Game/Screens/Play/Player.cs

```diff
@@ -748,39 +759,67 @@ public override bool OnExiting(IScreen next)
             return base.OnExiting(next);
         }
 
-        protected virtual void GotoRanking()
+        /// <summary>
+        /// Creates the player's <see cref="Score"/>.
+        /// </summary>
+        /// <returns>The <see cref="Score"/>.</returns>
+        protected virtual Score CreateScore()
         {
+            var score = new Score
+            {
+                ScoreInfo = new ScoreInfo
+                {
+                    Beatmap = Beatmap.Value.BeatmapInfo,
+                    Ruleset = rulesetInfo,
+                    Mods = Mods.Value.ToArray(),
+                }
+            };
+
             if (DrawableRuleset.ReplayScore != null)
             {
-                // if a replay is present, we likely don't want to import into the local database.
-                this.Push(CreateResults(CreateScore()));
-                return;
+                score.ScoreInfo.User = DrawableRuleset.ReplayScore.ScoreInfo?.User ?? new GuestUser();
+                score.Replay = DrawableRuleset.ReplayScore.Replay;
+            }
+            else
+            {
+                score.ScoreInfo.User = api.LocalUser.Value;
+                score.Replay = new Replay { Frames = recordingScore?.Replay.Frames.ToList() ?? new List<ReplayFrame>() };
             }
 
-            LegacyByteArrayReader replayReader = null;
+            ScoreProcessor.PopulateScore(score.ScoreInfo);
 
-            var score = new Score { ScoreInfo = CreateScore() };
+            return score;
+        }
 
-            if (recordingScore?.Replay.Frames.Count > 0)
-            {
-                score.Replay = recordingScore.Replay;
+        /// <summary>
+        /// Submits the player's <see cref="Score"/>.
+        /// </summary>
+        /// <param name="score">The <see cref="Score"/> to submit.</param>
+        /// <returns>The submitted score.</returns>
+        protected virtual async Task<ScoreInfo> SubmitScore(Score score)
```

Comment:
Bit confused by this being called `SubmitScore` but instead running import code. Might be worth moving the replay import code to its own method and having the base method empty here?

Reply:
Yeah I knew this was going to come up. I struggled with this too. So you're suggesting to split it out into `ImportScore` and `SubmitScore`?

Reply:
I've split it out into two methods. I like this structure more if only to not worry about ordering of `base` calls - `TimeshiftPlayer` previously awaited `base.SubmitScore()` at the very end to order the import process such that it has an online score id.